perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -5538,6 +5538,61 @@ public void addGroupsToAutoRegistration(PerunSession sess, List<Group> groups) {
 		}
 	}
 
+	@Override
+	public List<List<Group>> getIndirectMembershipPaths(PerunSession sess, Member member, Group group) throws MemberNotExistsException, GroupNotExistsException {
+		if (!isGroupMember(sess, group, member))
+			throw new MemberNotExistsException("Member does not belong to this group");
+
+		return stepIntoGroup(sess, member, group);
+	}
+
+	/**
+	 * Step into group, if it matches the source, start reconstructing the path, otherwise step into all
+	 * subgroups where member is recognised and glue current group to path when returning. Included groups
+	 * are cut off after first included group. Duplicates avoided.
+	 */
+	private List<List<Group>> stepIntoGroup(PerunSession sess, Member member, Group currentGroup) {
+		List<List<Group>> pathsFromCurrentGroup = new ArrayList<>();
+
+		// reached source group, create empty list to which current group will be added at the end of this method call
+		if (isDirectGroupMember(sess, currentGroup, member)) {
+			pathsFromCurrentGroup.add(new ArrayList<Group>());
+		}
+
+		// subgroups and included groups
+		List<Group> relatedGroups = getGroupUnions(sess, currentGroup, false);
+
+		// step into every group in which the member is recognised
+		for (Group group : relatedGroups) {
+			if (isGroupMember(sess, group, member)) {
+
+				// cut paths via included groups
+				if (!this.getGroupsManagerImpl().getSubGroups(sess, currentGroup).contains(group)) {
+					List<Group> cutPath = new ArrayList<>();
+					cutPath.add(group);
+					pathsFromCurrentGroup.add(cutPath);
+					continue;
+				}
+
+				List<List<Group>> subPaths = stepIntoGroup(sess, member, group);
+				pathsFromCurrentGroup.addAll(subPaths);
+			}
+		}
+
+		List<List<Group>> uniquePathsFromCurrentGroup = pathsFromCurrentGroup.stream()
```

Comment:
Is it really possible to have duplicate paths when we iterate through each related group just once?